### DISCUSSION TITLE
Advice for converting an ncurses program (lnav) to notcurses

### DISCUSSION THREAD

**Discussion by tstack** [dankamongmen/notcurses#2810] [Q&A]
I'm the author of [lnav](https://lnav.org) and I'm a bit tired of working around ncurses limitations and looking for something to replace it.  I've started the conversion to notcurses, but I would like some advice on applying styling to text.  lnav currently works by building up an "attributed line" data structure that contains the text to put on the screen and a collection of metadata for spans of that text.  The metadata can include visual stuff, like colors and styles, along with other stuff, like the file the log line came from.

Now, when rendering an attributed line to the ncurses API:

1. write the text with `waddnstr()`
1. read back the just written text into an array of `cchar_t` using `mvwin_wchnstr()`
1. resolve all the overlapping style attributes from the attributed line into the array of `cchar_t`s
   (an example of overlap is when there is a background color that applies to part of a line and inside of that section a foreground color change)
1. use `mvwadd_wchnstr()` to write back the `cchar_t`s

I arrived at this method because I ran into various issues trying to change the attributes as I went, using something like `wattr_set()`.

Looking through the notcurses API, it looks like I can sorta keep this method by writing the string and then operating on the `nccell`s.  However, it looks like I can only operate on a single cell at a time, I can't read/write an array of cells.  Is that right?  Or, is there a better way I should be structuring things.

**Comment by dankamongmen** [comment]
hey there! great question. i think i actually have something of a (very high level) document on this...let me look. Yeah, here's the doc: https://github.com/dankamongmen/notcurses/blob/master/doc/CURSES.md. Now let me look at your questions...

**Comment by dankamongmen** [comment]
ok, so you almost never need work with `nccell`s, at least not as your default interface with the library. if you are, you're probably doing more work than you need.

all the stuff about merging colors etc. you describe above is something you very much want to leave to Notcurses. overlapping texts ought be difference `ncplane`s, ordered to match reality. if you make the background `NCALPHA_TRANSPARENT`, it'll go to the next plane down to figure out the background. this is all described in `notcurses_render(3)` under the `Cell rendering algorithm` section.

check chapter 8 in [the book](https://nick-black.com/htp-notcurses.pdf): 

> It’s useful to note that there is never a need for more than one plane, as demonstrated by the simple fact that
each rendered frame is a two-dimensional area. Planes do not add power to the system: any algorithm which
can be expressed using multiple planes can be expressed using a single plane and external state. Instead, they
add expressiveness, and supply order to the state needed beyond the data present in the rendered frame.
The color blending performed for transparent or translucent planes can be simulated by the programmer.
Redrawing the parts of an underlying plane exposed by moving another can be managed by the programmer.
Mapping a base glyph to null cells of different planes can be done by keeping an index for each null cell, etc.
etc. In a great many cases, this external structure would be reproducing the algorithms and data structures
of planes. **Planes are provided as a concise and efficient implementation of the codes frequently
necessary to implement TUIs.** The rule for using planes is thus simply to *use a plane whenever you find
yourself implementing code already provided by planes*.

so all your reading-back-and-setting should be dropped. write to planes using the style you want. render the pile. it will look like you want it to, based on the ordering of the `ncplane`s within that `ncpile`.

does that make sense?

**Reply by tstack** [reply]
I think I understand the flow you're suggesting, but I'm concerned about having to manage all of these `ncplane` objects.  It looks like I can use the `ncplane_mergedown_simple()` function so I can create a styling plane, merge it down to the main plane I'm working with, and then destroy the styling plane.  Is that the intention?

**Reply by tstack** [reply]
With this approach, how does one do reverse video?  It looks like you removed `NCSTYLE_REVERSE`.  So, I can't just make a plane with just that style and merge it down.

**Comment by dankamongmen** [comment]
now looking at lnav, it looks like it is not a TUI per se? does it use the alternate screen, or at least disrupt what was there before? or does it only scroll along like e.g. `cat`?

**Reply by tstack** [reply]
> does it use the alternate screen

Yes, it uses the alternate screen.

> does it only scroll along like e.g. cat?

No, it's a viewer like more/less.  This [blog post](https://lnav.org/2023/06/23/cursor-mode.html) has an asciinema capture that shows what it looks like.
